test/soake2erekt/main_test.go

```diff
@@ -0,0 +1,254 @@
+The features are generated dynamically by a function, so that it is possible to generate unique resource names
+for each test copy, or in each iteration
+*/
+type SoakFeatureFn func(SoakEnv) *feature.Feature
+
+type SoakTest struct {
+	/*
+		Prefix to be used for namespaces. Actual test namespace is <prefix><copyId>
+	*/
+	NamespacePrefix string
+	/*
+		Function invoked during setup for each soak test copy. Use RunSoakFeature* inside to handle cleanup of the resources created by the features
+	*/
+	SetupFn SoakFn
+	/*
+		Function invoked for each iteration.
+		Use RunSoakFeature* inside to handle cleanup of the resources created by the features.
+		Use RunSoakFeatureFn* inside to also handle creation of the features dynamically based on the context of the soak test
+	*/
+	IterationFn SoakFn
+	/*
+		Function invoked after the iterations complete. Could be used to teardown or state verification after the end of the soak test.
+	*/
+	TeardownFn SoakFn
+}
+
+type soakKey struct{}
+
+type SoakEnv interface {
+	CopyID() int
+	Iteration() int
+	Namespace() string
+}
+
+type soakEnvImpl struct {
+	copyID    int
+	iteration int
+	namespace string
+}
+
+func (env *soakEnvImpl) CopyID() int {
+	return env.copyID
+}
+
+func (env *soakEnvImpl) Iteration() int {
+	return env.iteration
+}
+
+func (env *soakEnvImpl) Namespace() string {
+	return env.namespace
+}
+
+func soakEnvImplFromContext(ctx context.Context) *soakEnvImpl {
+	if se, ok := ctx.Value(soakKey{}).(*soakEnvImpl); ok {
+		return se
+	}
+	panic("no soak environment found in the context, make sure you're executing a soak test code within a RunSoakTest")
+}
+
+func SoakEnvFromContext(ctx context.Context) SoakEnv {
+	return soakEnvImplFromContext(ctx)
+}
+
+func RunSoakTest(t *testing.T, test SoakTest, copies int) {
+	for copyID := 0; copyID < copies; copyID++ {
+		copyID := copyID
+		namespace := test.NamespacePrefix + strconv.Itoa(copyID)
+		t.Run(namespace, func(t *testing.T) {
+			t.Parallel()
+
+			since := time.Now()
+
+			ctx, env := soakTestEnvironment(t, namespace)
+
+			// Execute the setup "features", store the references created during setup for cleanup at the end
+			setupCtx := context.WithValue(ctx, soakKey{}, &soakEnvImpl{
+				copyID:    copyID,
+				iteration: -1,
+				namespace: namespace,
+			})
+
+			if test.SetupFn != nil {
+				test.SetupFn(setupCtx, env, t)
+			}
+			if t.Failed() {
+				feature.LogReferences(env.References()...)(ctx, t)
+				return
+			}
+
+			iteration := 0
+			// Repeat the soak test for the duration specified by --soak-duration flag
+			for since.Add(Flags.Duration).After(time.Now()) {
+				// During each iteration, generate the "iteration" features and run them as Tests
+				// Cleanup all resources for these features at the end of the iteration
+
+				// TODO: can we do this? (use an empty env for the iteration, but keep using the original context?)
+				_, iterationEnv := global.Environment(
```

Comment:
If I understand it correctly, you create a new environment so as to be able to delete resources from the given iteration via `iterationEnv.References()` ? 
It should work but we're relying on some implementation details that might change in the future. 
Maybe it would be easier to just call `iterationCtx, iterationEnv := soakTestEnvironment(t, namespace)` here, and use the given iteration context and env in this iteration. Then you wouldn't need  to call `iterationCtx = environment.ContextWith(iterationCtx, iterationEnv)` because it's already done when creating the iterationCtx and iterationEnv

Reply:
we need the context to be per soak test, not per iteration, as things like eventshub use the context to store the eventshub store in it... (see `registerEventsHubStore`  ) (so, to be able to use eventshub receiver in soak setup, but use eventshub sender per iteration)

Reply:
Thanks. It was not immediately clear. I think we can do what you did. We just need to be extra careful which `ctx` or `env` is used later in the code.

Reply:
We probably also want to set some env here, maybe at least these three?:
```
knative.WithKnativeNamespace(system.Namespace()),
environment.WithPollTimings(Flags.PollInterval, Flags.PollDuration),
environment.WithTestLogger(t),
```